File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_cpp_gpt/3_multi_thread_basic.cpp
use std::sync::{Arc, Mutex};
use std::thread;

fn withdraw(balance: Arc<Mutex<i32>>) {
    let amount = 500000;
    for _ in 0..amount {
        let mut balance = balance.lock().unwrap();
        *balance -= 1;
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1000000));

    let balance_clone1 = Arc::clone(&account_balance);
    let t1 = thread::spawn(move || {
        withdraw(balance_clone1);
    });

    let balance_clone2 = Arc::clone(&account_balance);
    let t2 = thread::spawn(move || {
        withdraw(balance_clone2);
    });

    t1.join().unwrap();
    t2.join().unwrap();

    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", *final_balance);
}

File: ga/glm4_result_2024_6/result/multi_thread_basic_glm/3_multi_thread_basic.rs
use std::thread;
use std::sync::{Mutex, Arc};

fn main() {
    let account_balance = Arc::new(Mutex::new(1000000));  // Shared account balance with a mutex
    let mut handles = Vec::new();  // Vector to hold thread handles

    for _ in 0..2 {
        let account_balance = Arc::clone(&account_balance);  // Clone the shared account balance
        let handle = thread::spawn(move || {
            let mut balance = account_balance.lock().unwrap();  // Lock the mutex and get the balance
            for _ in 0..500000 {
                *balance -= 1;  // Decrement the balance
            }
        });
        handles.push(handle);  // Push the handle to the vector
    }

    for handle in handles {
        handle.join().unwrap();  // Wait for all threads to finish
    }

    println!("Final account balance: {}", *account_balance.lock().unwrap());  // Print the final balance

    // No need to explicitly destroy the mutex or Arc, they are automatically cleaned up when they go out of scope
}